src/VisualStudio/Core/Def/Copilot/VisualStudioCopilotCodeAnalysisService.cs

```diff
@@ -0,0 +1,215 @@
+        IThreadingContext threadingContext,
+        IVsService<SVsSettingsPersistenceManager, ISettingsManager> settingsManagerService)
+    {
+        _copilotDiagnosticUpdateSource = copilotDiagnosticUpdateSource;
+        _diagnosticsRefresher = diagnosticsRefresher;
+        _globalOptions = globalOptions;
+        InitializeAsync(settingsManagerService, threadingContext.DisposalToken).Forget();
+    }
+
+    private async Task InitializeAsync(IVsService<SVsSettingsPersistenceManager, ISettingsManager> settingsManagerService, CancellationToken cancellationToken)
+    {
+        _settingsManager = await settingsManagerService.GetValueAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    public bool IsRefineOptionEnabled(Document document)
+        => IsCopilotOptionEnabled(document, "EnableCSharpRefineQuickActionSuggestion");
+
+    public bool IsCodeAnalysisOptionEnabled(Document document)
+        => IsCopilotOptionEnabled(document, "EnableCSharpCodeAnalysis");
+
+    private bool IsCopilotOptionEnabled(Document document, string optionName)
+    {
+        // The bool setting is persisted as 0=None, 1=True, 2=False, so it needs to be retrieved as an int.
+        if (_settingsManager?.TryGetValue($"Microsoft.VisualStudio.Conversations.{optionName}", out int isEnabled) != GetValueResult.Success)
+            return false;
+
+        return isEnabled == 1 && document.GetLanguageService<ICopilotAnalyzer>() is not null;
+    }
+
+    private static bool IsLspPullDiagnostics(IGlobalOptionService globalOptions)
+        => globalOptions.IsLspPullDiagnostics(InternalDiagnosticsOptionsStorage.NormalDiagnosticMode);
+
+    public Task<bool> IsAvailableAsync(Document document, CancellationToken cancellationToken)
+    {
+        var analyzer = document.GetLanguageService<ICopilotAnalyzer>();
+        return analyzer is null
+            ? Task.FromResult(false)
+            : analyzer.IsAvailableAsync(cancellationToken);
+    }
+
+    public Task<ImmutableArray<string>> GetAvailablePromptTitlesAsync(Document document, CancellationToken cancellationToken)
+    {
+        if (!IsCodeAnalysisOptionEnabled(document))
+            return Task.FromResult(ImmutableArray<string>.Empty);
+
+        var analyzer = document.GetRequiredLanguageService<ICopilotAnalyzer>();
+        return analyzer.GetAvailablePromptTitlesAsync(document, cancellationToken);
+    }
+
+    private async Task<bool> ShouldSkipAnalysisAsync(Document document, CancellationToken cancellationToken)
+    {
+        if (!IsCodeAnalysisOptionEnabled(document))
+            return true;
+
+        if (await document.IsGeneratedCodeAsync(cancellationToken).ConfigureAwait(false))
+            return true;
+
+        return false;
+    }
+
+    public async Task AnalyzeDocumentAsync(Document document, TextSpan? span, string promptTitle, CancellationToken cancellationToken)
+    {
+        if (_executingAnalysis)
+            return;
+
+        _executingAnalysis = true;
```

Comment:
this is not threadsafe.  and our services should not be stateful in this fashion.